### DISCUSSION TITLE
Trying to persist Window Size and position

### DISCUSSION THREAD

**Discussion by David-H-Bolton** [tui-cs/Terminal.Gui#1610] [Q&A] [answered]
I've got a Window that opens and can save win.Frame.X, win.Frame.Y, win.Frame.Width and win.Frame.Height as strings into a text file. I use these values to create the Window size and that mostly works. Also in the Application.Resized event I change the Window Width and Height and that corrects the border. I.e. it matches the terminal window. 

But when I load the saved values on the next run after resizing, the border is wrong.
![border](https://user-images.githubusercontent.com/10566162/154952544-afea6240-cfe9-4cc4-97bb-bc1b5f4c4437.png)

I've used Application.Refresh() but that doesn't make a difference. So  am I just doing the wrong thing? Is there a better way to persist Window size and position?

**Comment by tznind** [comment]
If your window is `Width = Dim.Fill()`  (the default) then you want to persist the fact that its width is Fill not that it's width is 40 (or whatever).

There are 2 layouts `Computed` and `Absolute`.  If you are storing the current values of a `Computed` layout then depersisting that as `Absolute` values then it won't resize.

Ideally you would want to pass off persistence to a serialization library like JSON or YAML.

I looked into what happens when you pass `Window` to YAML Dot Net to serialize and it did not go well.  I spent an hour messing about with type converters but didn't make much headway.  I think it would take quite a lot of work to write a functioning serializer that works for all the right properties and skips the ones that reference other properties under the hood or otherwise behave unexpectedly.

**Reply by David-H-Bolton** [reply]
For the rest of the application I'd be happy to serialize state- I usually use JSON but for the top level Window, all I need are absolute positions and size. Why is it so hard to do?

**Reply by tznind** [reply]
Lets take a step back a second.  To make a `Window` that fills all (or a portion) of the screen and resizes automatically you should probably use computed layout not Application.Resized event (see below).  Then all you need to do is persist whether the screen is showing or not and construct it in the same way?

```csharp
var fullWindow = new Window("Test Window")
    {
        Width = Dim.Fill(),
        Height = Dim.Fill(),
    };

    var smallWindow = new Window("Test Window")
    {
        X = Pos.Percent(25),
        Y = Pos.Percent(25),
        Height = Dim.Percent(50),
        Width = Dim.Percent(50),
    };
```

**Comment by David-H-Bolton** [comment]
Am I right in thinking there's no way to control the Terminal Window size? I put in a Resized method and used Dim.Fill() to set Window after a resize and that worked but setting the terminal size in code doesn't seem possible.

**Reply by BDisp** [reply]
If you are talking about resizing the console window based on a saved state, that only possible if you didn't closed the console window yet and loading another toplevel with the `Application.Init`. See the `UICatalog` about the way it uses the `Application.Init`, `Application.Shutdown` and `Application.Init` again to start another scenario. It always respect the last window state. If you close the console window and relaunch the app will load with the default properties. Maybe it's possible to do that inside the `Terminal.Gui` but not for end users. The `ConsoleDriver`'s `Cols` and `Rows` properties are read only.

**Comment by David-H-Bolton** [comment]
When my application starts, before creating the main Window, it looks for the file with the previous values and if present reads those. It then creates the Window using the Width and Height using those.  That worked fine except you get the situation as the image above shows.

**Reply by tznind** [reply]
You only need to set the Width to `Dim.Fill()` once (when you create the Window).  Terminal.Gui will automatically resize for you.  Theres no need to use the `Resized` method at all.  `Dim.Fill()` doesn't just calculate the current value it also tracks future changes (e.g. when terminal is resized).

> setting the terminal size in code doesn't seem possible.

You are right that a terminal process cannot resize the terminal window itself.  That window (e.g. powershell) is under the control of the OS (e.g. Windows) not the hosted application.  Any solution would be platform specific (e.g. sending windows messages to the Powershell window).

**Comment by David-H-Bolton** [comment]
Sorry, I wasn't clear- the Resize method is the event handler that is run when the Terminal is manually resized. In it I set Width and Height to Dim.Fill() and that puts the Border in the correct place.

**Reply by tznind** [reply]
Are you able to share your current code?

**Comment by David-H-Bolton** [comment, marked as answer]
It's currently about 100 lines long. and I'm just trying to put together a basic skeleton. The main things were the Load and Save of the main Window. But those don't work. Anyway I've uploaded it in a zip file.
[Lib.zip](https://github.com/migueldeicaza/gui.cs/files/8110496/Lib.zip)

**Reply by tznind** [reply]
Thanks for sharing the code.  I have cut out the load and save and shown how to make a `Window` automatically resize without creating any event handlers etc.  You can see that after closing and reopening the application (see the video gif) it still resizes.

I don't understand why you are loading and saving the window coordinates when they are auto computed by the system already (when using `Dim.Fill`).  If you want multiple windows at once then happy to write a skeleton for that too (or an arbitrary number of windows).


```csharp
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Terminal.Gui;

namespace CTL
{
    internal class Lib
    {
        static Window? win = null;

        public static void Main(string[] args)
        {
            Application.Init();
            var menu = new MenuBar(new MenuBarItem[] {
                new MenuBarItem ("_File", new MenuItem [] {
                    new MenuItem ("_Quit", "", () => {
                        Application.RequestStop ();
                    })
                })
            });

            win = new Window("test")
            {
                X = 0,
                Y = 1,
                Width = Dim.Fill(),
                Height = Dim.Fill(1)
            };            
            win.ColorScheme = Colors.TopLevel;
            Application.Top.Add(menu, win);
            
            Application.Run();
            
            Application.Shutdown();
        }
    }
}
```

![resizing](https://user-images.githubusercontent.com/31306100/154991483-95b83697-1697-42f0-9c60-1e31c0896a43.gif)

**Reply by David-H-Bolton** [reply]
I think its my lack of experience with gui.cs. I'm glad that these discussions exist.  Interesting, I see you are using **dotnet run** on Linux?. When I run this on Windows in Visual Studio (2022) or in a command line it doesn't save position or size.

**Reply by BDisp** [reply]
Because the Visual Studio is always opening a new console window. If you open on a console (cmd, powershell or terminal windows) and do a dotnet run, do a resize, close the app and do again dotnet run. You will see it will respect the console window size.
In these case you don't need to save the pos state.

**Reply by David-H-Bolton** [reply]
Yes, just tried it and confirm that it derives its size from the console window. That's sorted then! Thank you.

**Reply by JessicaMulein** [reply]
Love seeing the community interaction and the help. This is what makes open source amazing.
